include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,423 @@
+                    const double distributionIndex,
+                    const arma::vec& lowerBound,
+                    const arma::vec& upperBound) :
+    populationSize(populationSize),
+    numGeneration(numGeneration),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    neighbourhoodSize(neighbourhoodSize),
+    distributionIndex(distributionIndex),
+    lowerBound(lowerBound),
+    upperBound(upperBound),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+inline MOEAD::MOEAD(const size_t populationSize,
+                    const size_t numGeneration,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const size_t neighbourhoodSize,
+                    const double distributionIndex,
+                    const double lowerBound,
+                    const double upperBound) :
+    populationSize(populationSize),
+    numGeneration(numGeneration),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    neighbourhoodSize(neighbourhoodSize),
+    distributionIndex(distributionIndex),
+    lowerBound(lowerBound * arma::ones(1, 1)),
+    upperBound(upperBound * arma::ones(1, 1)),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type MOEAD::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Check if lower bound is a vector of a single dimension.
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
+  if (upperBound.n_rows == 1)
+    upperBound = upperBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+  std::vector<size_t> shuffle(populationSize);
```

Comment:
I think we can use the `shuffel` method thats part of the function implementation, see https://github.com/mlpack/ensmallen/blob/60a99f780ba24fb0403579534bc53db9e09f54c9/include/ensmallen_bits/sgd/sgd_impl.hpp#L200-L201 for an example.

Reply:
Should I add a Shuffle function to problem or use arma::shuffle and arma::linspace as used in the Shuffle function above?

Reply:
Personally I would rely on the shuffel function that is part of the problem, for example:

https://github.com/mlpack/mlpack/blob/e11c3e320c4008a3c6ed16abeef69d53c0584366/src/mlpack/methods/ann/ffn.hpp#L282

implement the `Shuffel` function

https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/logistic_regression/logistic_regression_function.hpp#L57

as well, basically all methods in mlpack, so I think it makes sense to reuse that functionality.

Reply:
@zoq, since there is more than one objective, and we have access to the object of each objective via the tuple, I have implemented Shuffle() inside the first problem. Is this fine? I was wondering, isn't Shuffle() in SGDTestFunction used to change the order in which the three components of the function are added, i.e. wasn't it affecting the problem? The shuffle in our case is affecting the optimizer itself and not the problem, so I am still not sure if this is correct.

Reply:
@zoq, can you review this please? I thought I would update the docs once everything is finalised.

Reply:
`SGDTestFunction` that is correct, but I'm almost sure that's what was implemented but was part of the optimizer; will double check to make sure this is still correct.